src/main/java/org/eclipse/microprofile/starter/log/DynamoDBLogger.java

```diff
@@ -0,0 +1,318 @@
+    private static final String service = "dynamodb";
+    private static final String host = service + "." + region + ".amazonaws.com";
+    private static final String endpoint = "https://" + service + "." + region + ".amazonaws.com/";
+    private static final String contentType = "application/x-amz-json-1.0";
+    private static final String amzTarget = "DynamoDB_20120810.PutItem";
+    private static final String algorithm = "AWS4-HMAC-SHA256";
+    private static final String signedHeaders = "content-type;host;x-amz-date;x-amz-target";
+
+    /**
+     * Return value for the logger
+     */
+    public enum Status {
+        OK,
+        NOK,
+        NOT_CONFIGURED
+    }
+
+    private static byte[] sign(final byte[] key, final byte[] message) throws NoSuchAlgorithmException, InvalidKeyException {
+        final Mac mac = Mac.getInstance("HmacSHA256");
+        mac.init(new SecretKeySpec(key, "HmacSHA256"));
+        return mac.doFinal(message);
+    }
+
+    private static byte[] signatureKey(
+            final String key, final String date_stamp, final String regionName, final String serviceName) throws NoSuchAlgorithmException, InvalidKeyException {
+        final byte[] kDate = sign(StandardCharsets.UTF_8.encode("AWS4" + key).array(), StandardCharsets.UTF_8.encode(date_stamp).array());
+        final byte[] kRegion = sign(kDate, StandardCharsets.US_ASCII.encode(regionName).array());
+        final byte[] kService = sign(kRegion, StandardCharsets.US_ASCII.encode(serviceName).array());
+        return sign(kService, StandardCharsets.US_ASCII.encode("aws4_request").array());
+    }
+
+    private static String preparePayload(final EngineData engineData, final Calendar calendar) {
+        final SimpleDateFormat logMessageTimeFormat = new SimpleDateFormat(logRecordTimestampFormat);
+        final MessageDigest sha1Hash;
+        try {
+            sha1Hash = MessageDigest.getInstance("SHA-1");
+        } catch (NoSuchAlgorithmException e) {
+            Log.log(Level.SEVERE, "SHA-1 not available.", e);
+            return null;
+        }
+        final String timestamp = logMessageTimeFormat.format(calendar.getTime());
+        final String logmark = Stream.concat(
+                Stream.of(engineData.getMpVersion(), engineData.getSupportedServer(), engineData.getBeansxmlMode(), webAppInstanceID,
+                        timestamp, engineData.getMavenData().getArtifactId(), engineData.getMavenData().getGroupId()),
+                engineData.getSelectedSpecs().stream()
+        ).collect(Collectors.joining());
+        sha1Hash.update(logmark.getBytes());
+        final String hashedLogmark = DatatypeConverter.printHexBinary(sha1Hash.digest());
+
+        final StringBuilder s = new StringBuilder(1024)
+                .append("{\"TableName\":\"")
+                .append(tableName)
+                .append("\",\"Item\":{")
+                .append("\"logmark\":{\"S\":\"")
+                .append(hashedLogmark)
```

Comment:
Is this something as a Primary Key of the table? Because it will not be unique.

Reply:
@rdebusscher As I commented earlier:

 > Each row has to be uniquely identified (this is a db that does not do any autoincrement
 > something by design as it requires relation between records). Logmark is a hash of
 > timestamp and logged data. If anybody manages to send more same download requests
 > within the same millisecond the row gets collision and gets overwritten. I deem this not a
 > problem because numerous downloads with exactly same parameters including artifact and
 > group id within a millisecond is an abuse and should not be logged as a legitimate download
 > anyway.

If we want to track exactly all downloads, including those abusing the service, we would have to take different approach or shard-distributed locking, retrieving latest value, incrementing, writing, releasing lock...  This is not exactly the friendliest service for that, this is not a relational database per se, this is a sort of  NoSQL / document thing.

So my thesis is: It log record has this hash either unique or is is not a legitimate downlaod anyway.  If I am wrong, I can source more entrophy a lower the risk of a collision.

Or I would have to figure out how to get rid of that column altogether. It was not apparent to me from the AWS tutorial and docs that I could get about without such a special identifier per row.